src/Runtime/Loop.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+/*
+ * This file is part of Twig.
+ *
+ * (c) Fabien Potencier
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Twig\Runtime;
+
+use Twig\Error\RuntimeError;
+
+/**
+ * Represents a for loop variable.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+class Loop implements \Iterator
+{
+    private \Iterator $seq;
+    private bool $iterated;
+    private int $index0;
+    private int $revindex0;
+    private bool $first;
+    private bool $last;
+    private int $length;
+
+    public function __construct($seq, private $parent)
+    {
+        $this->seq = is_iterable($seq) ? (is_array($seq) ? new \ArrayIterator($seq) : $seq) : new \ArrayIterator([]);
+        $this->rewind();
```

Comment:
wouldn't this cause issues with a Generator because we would rewind the iterator twice (once in the constructor and once when the `foreach` loop calls the `rewind` method) while generators are not rewindable ?

We should have tests iterating over a Generator in Twig.

Reply:
I've just added a test with a Generator.